// autoprefixer-brunch/lib/index.js
// Generated by CoffeeScript 1.6.3
(function() {
  var AutoprefixerCompiler, autoprefixer;

  autoprefixer = require("autoprefixer");

  module.exports = AutoprefixerCompiler = (function() {
    AutoprefixerCompiler.prototype.brunchPlugin = true;

    AutoprefixerCompiler.prototype.type = "stylesheet";

    AutoprefixerCompiler.prototype.extension = "css";

    AutoprefixerCompiler.prototype.pattern = /\.(?:css|scss|sass|less|styl)$/;

    function AutoprefixerCompiler(config) {
      var _ref;
      this.config = config;
      this.browsers = ((_ref = this.config.plugins.autoprefixer) != null ? _ref : {}).browsers;
    }

    AutoprefixerCompiler.prototype.compile = function(_arg, callback) {
      var compiled, data, error, map, path, _ref;
      data = _arg.data, path = _arg.path, map = _arg.map;
      try {
        compiled = autoprefixer.apply(null, this.browsers).compile(data, {
          sourcemap: true,
          source: path
        });
      } catch (_error) {
        error = _error;
      }
      return callback(error, {
        data: (_ref = compiled != null ? compiled.code : void 0) != null ? _ref : compiled,
        map: compiled != null ? compiled.map : void 0
      });
    };

    return AutoprefixerCompiler;

  })();

}).call(this);
